### DISCUSSION TITLE
Native -> CLI -> Managed dll leading to EEFileLoadException

### DISCUSSION THREAD

**Discussion by DerekSMorin** [dotnet/runtime#70736] [Q&A]
I'm hitting the same file load exception in two different scenarios.

1.  CppUnitTest calls into mixed cli dll which then attempts to load a custom managed dll in the same folder and fails to load dll with EEFileLoadException.
2. Native console application calls into mixed cli dll which then attempts to load a custom managed dll in the same folder and fails to load dll with EEFileLoadException.

I wasn't sure if any of this was related to
https://github.com/dotnet/runtime/issues/61105
C++/CLI assembly is loaded into isolated AssemblyLoadContext (IsolatedComponentLoadContext) in .NET 6

I have also tried this in .NET preview 7.

My mixed cli dll is called oqkern.

If I call into it, and then access a standard .net api things work.  E.g. something that just works with managed strings or DateTime etc.

But if I call into it and it tries to access stuff from another custom c# dll in the same folder (e.g. Bentley.APM.Settings ) then if fails to load this dll.

Under .NET Preview 7 I am able to start debugging into the assembly binder code and I can see that in BindAssemblyByProbingPaths that pBindingPaths is an empty list.

![image](https://user-images.githubusercontent.com/7568856/173620219-62d38b51-118d-4532-9683-0948b819a8a5.png)

I'm wondering if what I am trying to do is supported or what the best way for me to achieve my goals is.

I'm not sure if I have to get into something complicated like:
Write a custom .NET host to control the .NET runtime from your native code
https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting

Or if I'm just missing something obvious.

I'll mention @vitek-karas and @elinor-fung as you both seemed very active in the discussions around the c++/cli assembly load context.

Thanks for your time,
Derek

**Comment by vitek-karas** [comment]
.NET Runtime (as oppose to .NET Framework runtime) does not probe for files on disk. It relies on `.deps.json` file which is produced by the SDK to determine the list of assemblies it should know about (where to find them). So this means that just copying a `.dll` next to another one doesn't make it discoverable by the runtime itself.

This is further complicated by the fact that you're loading the runtime from C++/CLI first. In .NET 6 the default behavior is to load only the core frameworks into the default load context, the C++/CLI assembly and all of its non-framework dependencies are loaded into a secondary load context. So in this case the default load context can only resolve assemblies from the framework (So basically `System.*` and nothing else).

In .NET 7 we changed this default and in this case the C++/CLI is loaded into the default load context which means that the default context will be able to resolve all of framework, the C++/CLI assembly and its dependencies. The dependencies of the C++/CLI assembly are determined by loading the `.deps.json` file next to it.

So if the assembly you're trying to load was not part of the component during build (typically either `ProjectReference` or `PackageReference`) then the runtime will not resolve it.

You can customize this via several different callbacks the runtime provides, the most typical for the default load context would be to add an event handler to the `AssemblyLoadContext.Default.Resolving` event. Here you can implement for example the policy to look for the assembly file next to the main dll, or whatever other policy makes sense in your case.

Detailed description of the assembly resolution algorithm can be found here: https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-managed (there are other related docs near it in the TOC as well).

**Reply by DerekSMorin** [reply]
Thanks @vitek-karas.  This was very informative and helpful.  Under .NET 6 it took me a while but I managed to get the "net-core hosting" sample going and then I still hit an EEFileLoadException.  I then added an assembly resolve in that code and I was able to get things going.
Then I tested just setting up an AssemblyResolve method in the c++/cli dll and I was able to get that going as well.  All of this was in the context of an executable.

I was also trying to get things going in the context of a CppUnitTest which was my actual goal.

I don't think anything I tried seemed to be working.  I'm wondering if it is related to CppUnitTest using .NET framework under the covers?

When I ran the unit test I noticed output:

'testhost.exe' (Win32): Loaded 'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\mscoreei.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\shlwapi.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\kernel.appcore.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140_clr0400.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbase_clr0400.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\psapi.dll'. Symbols loaded.
'testhost.exe' (Win32): Loaded 'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\NativeImages\mscorlib.ni.dll'. 
'testhost.exe' (Win32): Loaded 'C:\Windows\System32\bcryptprimitives.dll'. Symbols loaded.

Whereas in my executable I had logging:

'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\netapi32.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Development\Source\ApmUpdateDotNet6Loader\bin\x64\Debug-Net\net6.0\Ijwhost.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\uxtheme.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\dpapi.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\sspicli.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\netutils.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.1110_none_792d1c772443f647\comctl32.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\umppc15103.dll'. 
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Program Files\dotnet\host\fxr\7.0.0-preview.4.22229.4\hostfxr.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.5\hostpolicy.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.5\coreclr.dll'. Symbols loaded.
'Bentley.OQ.ExecutionContext.Console.exe' (Win32): Loaded 'C:\Windows\System32\bcryptprimitives.dll'. Symbols loaded.

We are wanting to write c++ tests against the c++/cli dll but currently they fail if it hits any "non-system" ( String^, DateTime^ ) managed code.  E.g. Some managed dlls created by us.

I don't know if there is any way around this and there is a potential we are just barking up the wrong tree.  Eventually we are wanting to get rid of the managed stuff in our mixed c++/cli dll and just have native code in there.  At that point things will probably go much more smoothly.  In the interim, we were hoping to be able to write CppUnitTests, but I'm not sure if there is a way we can do that.

Thanks again,

Derek

**Reply by vitek-karas** [reply]
I know very little about CppUnitTests - but in general .NET Framework and .NET (Core) doesn't interoperate (we don't really support running them side-by-side in the same process either). That is especially true for C++/CLI. Which .NET will be used is determined at build time, as C++/CLI project will generate different output when built for .NET Framework and .NET. If the CppUnitTest is a C++/CLI project and is built for .NET Framework, then it will run with .NET Framework. I don't know what will happen if you call a .NET C++/CLI from it ... it could get weird probably.

Instead of the ALC.Default.Resolving event you might be able to use AppDomain.AssemblyResolve event - that should work more or less the same way in .NET Framework and .NET.

**Reply by lvde0** [reply]
@vitek-karas Hi, I am sorry to revive this issue, but I am seeing a very similar problem as OP: I am calling a C++/CLI dll from native C++. The CLI dll depends on nuget libaries internally (included via `<PackageReference>`). These are correctly listed in the *.deps.json file next to the dll. However even with .NET 7 it still fails to load them at runtime (`EEFileLoadException`/`FileNotFoundException`). `ijwhost.dll` is loaded with version "7,0,2024,26716". It works when I manually copy the managed dlls in the bin folder, but I would like to avoid that for obvious reasons.

Is there anything else to consider?

EDIT: I checked the ALC of the CLI project and it's  `DefaultAssemblyLoadContext`, however the nuget dependencies are missing from the assemblies.

**Reply by vitek-karas** [reply]
@elinor-fung 

You can try https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#corehost_trace to get detailed logs from the dependency resolution (assuming it's due to the `.deps.json` file).
Where do you want to load the dependencies from, if they're not in the bin folder?

**Reply by lvde0** [reply]
Ideally, they would be loaded from the nuget packages folder/cache. At least this is how C#/managed projects usually behave.

To clarify my scenario I created a small repro here: 
[Sandbox.zip](https://github.com/user-attachments/files/17634132/Sandbox.zip)

**Reply by vitek-karas** [reply]
Ever since .NET 6 managed projects don't load dependencies from nuget cache - the default is to copy all dependencies into the apps directory (including nuget dependencies). Before .NET 6 the only way we made the nuget cache work was through `.runtimeconfig.dev.json` which would add the nuget cache to the probing paths. This only works on the machine where the project was built, and that was causing problems, so we removed that.

The case on C++/CLI is a bit more complicated as it may be loaded into a process which doesn't .NET in it yet - so those should carry their dependencies next to them for sure.

**Reply by lvde0** [reply]
Okay, I managed to get it working now by forcing the package artifacts to be copied to the output folder via `<CopyLocalLockFileAssemblies>`

**Reply by vitek-karas** [reply]
Glad to hear that it works for you. Alternatively, you may use https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#enabledynamicloading which is a superset and might be a better fit for this case.
